### ISSUE TITLE
Implement aspect

### ISSUE THREAD

**Issue by wkearn** [TopoToolbox/pytopotoolbox#109]
This is a useful function to have. We need it to complete the Getting Started guide section "Calculate basic DEM attributes."

There is an outstanding libtopotoolbox issue for aspect (TopoToolbox/libtopotoolbox#102). After the experiments that led to #100, we might consider a separate implementation here.

**Comment by Teschl** [comment]
Implementing this is proving harder than I imagined. As of right now, I have an implementation which produces plausible results, but does not replicate the Matlab function.

```python
grad_y, grad_x = np.gradient(self.z)
aspect = np.arctan2(-grad_x, grad_y)
aspect = np.degrees(aspect)
aspect = np.mod(aspect, 360)
```

I have not been able to find functions to be able to replicate the Matlab implementation. Neither the surfnorm function nor the cart2pol function seems to have a direct equivalent in any python package. At least, I could not find any. 

```matlab
[Nx,Ny] = surfnorm(Z);
ASP   = cart2pol(Nx,Ny);
ASP   = mod(90+ASP/pi*180,360);
```

### For Example:

Given the matrix:
```python
a = np.array([
    [1, 1, 1, 1],
    [1, 1, 1, 1],
    [1, 5, 2, 1],
    [1, 4, 3, 1]])
```
Produces the following result in python when aspect() is applied:
```
[[  0.           0.           0.           0.        ]
 [  0.           0.           0.           0.        ]
 [270.         341.56505118  63.43494882  90.        ]
 [270.         225.          56.30993247  90.        ]]
```

But in Matlab the result of the aspect function looks different:
```
90	180	180	90
90	0	0	90
270	341.565051177078	63.4349488229220	90
270	195.945395900923	56.3099324740202	90
```
I assume the differences can be attributed to different handling of the edges, although the two cells in the middle ate the bottom of the matrix are inconsistent, since one of the values matches in the implementations.

How should I handle this @wkearn ?

**Comment by wkearn** [comment]
Use `edge_order=2` in `np.gradient`. MATLAB's `surfnorm` uses a second order finite difference method, including at the edges.

`surfnorm` computes the gradients using the same formula as `np.gradient` with `edge_order=2` and then returns the components of the surface normal vector:

```python
nx = -grad_x / np.sqrt(grad_x ** 2 + grad_y**2 + 1)
ny = -grad_y / np.sqrt(grad_x ** 2 + grad_y**2 + 1)
nz = 1 / np.sqrt(grad_x ** 2 + grad_y**2 + 1)
```

Note the - sign in front of the x and y components. Then `cart2pol(nx,ny)` does `atan2(ny,nx)`. Then we convert to degrees, add 90 and use mod to get the angle into the 0,360 range.

Your solution works, though, because while you don't negate the `grad_y` component, you then leave off the offset by 90 degrees. It's just using a different coordinate system.

To summarize:

```python
grad_y, grad_x = np.gradient(self.z, edge_order=2)
aspect = np.arctan2(-grad_x, grad_y)
aspect = np.mod(np.degrees(aspect),360)
```

should work except for one minor detail:

When both `grad_y` and `grad_x` are zero. MATLAB's `atan2` function always returns 0, while `np.arctan2` chooses +0, -0, +180, or -180 depending on the signs of the arguments. 

When `grad_y` and `grad_x` are both zero, the surface is flat, and the aspect is not well defined anyway, so we might just have to live with this difference.

**Comment by Teschl** [comment]
I added the `edge_order=2` argument, but some issues at the edges still persist. I also implemented it using the `surfnorm` replacement you suggested, but the issues at the edges were not solved by that. It also did take two to three times longer to execute, so I suggest we should stick to the solution I used before.

Regarding the issues at the edges, still based on the following example:
```python
a = np.array([
    [1, 1, 1, 1],
    [1, 1, 1, 1],
    [1, 5, 2, 1],
    [1, 4, 3, 1]])
```

Now, the value that should be `90` in the bottom right is either `270` or `0`

```
Using new python implementation
[[270.         180.         180.         270.        ]
 [270.           0.           0.         270.        ]
 [270.         341.56505118  63.43494882 270.        ]
 [270.         195.9453959   56.30993247  90.        ]]

Using original python implementation
[[  0.         180.         180.           0.        ]
 [  0.           0.           0.           0.        ]
 [270.         341.56505118  63.43494882   0.        ]
 [270.         195.9453959   56.30993247  90.        ]]

Correct Matlab solution
90    180    180    90
90      0      0    90
270  341.565051177078  63.4349488229220   90
270  195.945395900923  56.3099324740202   90
```

@wkearn : I will create a Pull Request next, since I don't think there is much we can change about this weird behavior at the edges.

**Comment by wkearn** [comment]
I am pretty sure this is because of the issue with the signs of the zeros that I was talking about. Where the x and y gradients are both zero, numpy and MATLAB will produce different answers depending on the signs of the floating point zeros that you pass to `arctan2` ([This table](https://numpy.org/doc/stable/reference/generated/numpy.arctan2.html#id1) shows you the behavior in numpy. There is a similar table in the MATLAB documentation for `atan2`). These pixels just happen to be on the boundary in your example, but if you have flat regions on the interior of a DEM, you'll see the same thing. We can't really do anything about it, though, so we should just make a note in case we or someone else observes this difference in the future.
